jerry-core/parser/js/js-parser-expr.c

```diff
@@ -666,6 +666,63 @@ parser_parse_object_literal (parser_context_t *context_p) /**< context */
                                          literal_index,
                                          PARSER_OBJECT_PROPERTY_VALUE);
 
+#ifndef CONFIG_DISABLE_ES2015_SHORTHAND_OBJ_LITERAL_NOTATION
+      if (lexer_check_colon (context_p))
+      {
+        /* If there's a colon ahead, then we have { prop: val } */
+        lexer_next_token (context_p); /* Skip colon. */
+        lexer_next_token (context_p);
+        parser_parse_expression (context_p, PARSE_EXPR_NO_COMMA);
+
+        if (context_p->last_cbc_opcode == CBC_PUSH_LITERAL)
+        {
+          context_p->last_cbc_opcode = CBC_SET_LITERAL_PROPERTY;
+          context_p->last_cbc.value = literal_index;
+        }
+        else
+        {
+          parser_emit_cbc_literal (context_p, CBC_SET_PROPERTY, literal_index);
+        }
+      }
+      else if (context_p->token.lit_location.type == LEXER_IDENT_LITERAL)
+      {
+        /* If we have an identifier literal, then we have { prop } or { prop() {...} }
+           (a.k.a shorthand notation) */
+        if (lexer_check_left_paren (context_p))
+        {
+          parser_flush_cbc (context_p);
+          uint32_t status_flags = PARSER_IS_FUNCTION | PARSER_IS_FUNC_EXPRESSION | PARSER_IS_CLOSURE;
+          uint16_t function_literal_index = lexer_construct_function_object (context_p, status_flags);
+          parser_emit_cbc_literal (context_p, CBC_PUSH_LITERAL, function_literal_index);
+          lexer_next_token (context_p);
+        }
+        else
+        {
+          lexer_construct_literal_object (context_p,
+                                          &context_p->token.lit_location,
+                                          context_p->token.lit_location.type);
+          lexer_next_token (context_p);
+
+          /* We have a literal, but there is ill-formed syntax after it or it's a keyword,
+             so just error out. */
+          if ((context_p->token.type != LEXER_RIGHT_BRACE && context_p->token.type != LEXER_COMMA)
+              || lexer_is_identifier_keyword (context_p))
+          {
+            parser_raise_error (context_p, PARSER_ERR_COLON_EXPECTED);
+          }
+
+          parser_emit_cbc_literal_from_token (context_p, CBC_PUSH_LITERAL);
+        }
+
+        context_p->last_cbc_opcode = CBC_SET_LITERAL_PROPERTY;
+        context_p->last_cbc.value = literal_index;
+      }
+      else
+      {
+        /* Else, we have something that isn't a valid object literal. */
+        parser_raise_error (context_p, PARSER_ERR_COLON_EXPECTED);
+      }
```

Comment:
I don't really like this implementation. It is chaotic. You probably need a helper such as lexer_check_arrow, and add a switch.